src/gui/editlocallyhandler.cpp

```diff
@@ -0,0 +1,333 @@
+            showError(tr("Could not find a file for local editing. Make sure it is not excluded via selective sync."), _relPath);
+        } else {
+            showError(tr("Could not find a file for local editing. Make sure its path is valid and it is synced locally."), _relPath);
+        }
+        return;
+    }
+
+    _localFilePath = foundFiles.first();
+    _folderForFile = FolderMan::instance()->folderForPath(_localFilePath);
+
+    if (!_folderForFile) {
+        showError(tr("Could not find a folder to sync."), _relPath);
+        return;
+    }
+
+    const auto relPathSplit = _relPath.split(QLatin1Char('/'));
+    if (relPathSplit.isEmpty()) {
+        showError(tr("Could not find a file for local editing. Make sure its path is valid and it is synced locally."), _relPath);
+        return;
+    }
+
+    _fileName = relPathSplit.last();
+
+    Systray::instance()->destroyEditFileLocallyLoadingDialog();
+    Q_EMIT setupFinished();
+}
+
+QString EditLocallyHandler::prefixSlashToPath(const QString &path)
+{
+    auto slashPrefixedPath = path;
+    if (!slashPrefixedPath.startsWith('/')) {
+        slashPrefixedPath.prepend('/');
+    }
+
+    return slashPrefixedPath;
+}
+
+bool EditLocallyHandler::isTokenValid(const QString &token)
+{
+    if (token.isEmpty()) {
+        return false;
+    }
+
+    // Token is an alphanumeric string 128 chars long.
+    // Ensure that is what we received and what we are sending to the server.
+    const QRegularExpression tokenRegex("^[a-zA-Z0-9]{128}$");
+    const auto regexMatch = tokenRegex.match(token);
+
+    // Means invalid token type received, be cautious with bad token
+    if(!regexMatch.hasMatch()) {
+        return false;
+    }
+
+    return true;
+}
+
+bool EditLocallyHandler::isRelPathValid(const QString &relPath)
+{
+    if (relPath.isEmpty()) {
+        return false;
+    }
+
+    // We want to check that the path is canonical and not relative
+    // (i.e. that it doesn't contain ../../) but we always receive
+    // a relative path, so let's make it absolute by prepending a
+    // slash
+    const auto slashPrefixedPath = prefixSlashToPath(relPath);
+
+    // Let's check that the filepath is canonical, and that the request
+    // contains no funny behaviour regarding paths
+    const auto cleanedPath = QDir::cleanPath(slashPrefixedPath);
+
+    if (cleanedPath != slashPrefixedPath) {
+        return false;
+    }
+
+    return true;
+}
+
+bool EditLocallyHandler::isRelPathExcluded(const QString &relPath)
+{
+    if (relPath.isEmpty()) {
+        return true;
+    }
+
+    const auto folderMap = FolderMan::instance()->map();
+    for (const auto &folder : folderMap) {
+        bool result = false;
+        const auto excludedThroughSelectiveSync = folder->journalDb()->getSelectiveSyncList(SyncJournalDb::SelectiveSyncBlackList, &result);
+        for (const auto &excludedPath : excludedThroughSelectiveSync) {
+            if (relPath.startsWith(excludedPath)) {
```

Comment:
I am a bit confused. Should we not return `true` since `relPath` starts with `excludedPath` in `SelectiveSyncBlackList`?

Reply:
I mixed up my bools :)